### ISSUE TITLE
How to use a GitHub branch or commit hash instead of an npm package version?

### ISSUE THREAD

**Issue by jaydenseric** [jspm/project#152]
How can a GitHub branch or specific commit be specified in an import map for a project leveraging the JSPM CDN to test a not yet published version or fork of a dependency?

In my specific case, I would like to install the `preact` `main` branch to workaround rendering bugs that will be fixed when they publish v11:

https://github.com/preactjs/preact/issues/2783#issuecomment-993028422

This is only intended as a temporary thing to unblock development; hopefully `preact` v11 is out before I'm ready to publish my dependant package and related projects.

**Comment by guybedford** [comment]
I would generally recommend a local cloning workflow here, especially given many GitHub packages require a build step anyway.

To point the resolution of preact to a locally cloned directory, you can use the "resolutions" option in the generator, see https://github.com/jspm/generator#resolutions which allows pointing a package to a local directory and the resolutions will follow the package.json resolutions of the package further.

**Comment by jaydenseric** [comment]
Sorry, from that description I'm struggling to imagine how it all fits to together. A high level step by step guide would be really helpfull.

Trying out forks in projects is something really common; in time there will help a lot of other people in a similar predicament that could benefit from a guide or some docs.

**Comment by jaydenseric** [comment]
I'm a Deno user, so to get a local Deno script to output an import map using the local `preact` clone, first I need to generate an import map to use `@jspm/generator` in the script. Unfortunately, it results in this error:

> TypeError: Module specifier, 'to-fast-properties' does not start with "/", "./", or "../". Referenced from https://ga.jspm.io/npm:@babel/types@7.15.0/_/285570bb.js

https://generator.jspm.io/#a+JhYGBkDM0rySzJSU1hcMgqLsjVT0/NSy1KLMkvcjDUM9Az0E1KLUnUM7QEAFlDSfQsAA

It seems Deno people that want to quickly try a fork in their project will have to install Node just to use `@jspm/generator`.

**Comment by jaydenseric** [comment]
It doesn't seem to work?

<img width="1087" alt="Screen Shot 2021-12-14 at 7 38 54 pm" src="https://user-images.githubusercontent.com/1754873/145963034-4ba9eacd-c8f5-41b0-803a-9edc78b56ac7.png">

There is no error message if you make up a non existent directory file URL in the `resolutions` field, does that mean it's not even being used?

**Comment by guybedford** [comment]
Rather than installing preact, just generate the import map for your application that uses preact, then the resolutions will be used.

Alternatively install the folder itself directly if you want to just install preact as a local folder (`{ target: './preact', name: 'preact' }` as the install object.

Agreed resolutions should apply there, it's an edge case I hadn't thought of.

**Comment by guybedford** [comment]
@jaydenseric we have a Deno self-install in the tests so I'd be quite surprised if the Deno map isn't generating. See https://github.com/jspm/generator/blob/main/test/deno/self.test.js. Are you definitely setting `env: ['production', 'node', 'deno', 'module', 'test']` when generating for Deno?

**Comment by jaydenseric** [comment]
> Are you definitely setting `env: ['production', 'node', 'deno', 'module']` when generating for Deno?

I shared the link, so you can see it for yourself:

https://generator.jspm.io/#a+JhYGBkDM0rySzJSU1hcMgqLsjVT0/NSy1KLMkvcjDUM9Az0E1KLUnUM7QEAFlDSfQsAA


This is all very confusing. So here is my apps normal `importMap.json`:

```json
{
  "imports": {
    "class-name-prop": "https://ga.jspm.io/npm:class-name-prop@4.0.0/classNameProp.mjs",
    "device-agnostic-ui/": "https://ga.jspm.io/npm:device-agnostic-ui@9.0.0/",
    "dotenv/": "https://deno.land/x/dotenv@v3.0.0/",
    "graphql-react/": "https://ga.jspm.io/npm:graphql-react@16.0.0/",
    "marked": "https://ga.jspm.io/npm:marked@4.0.5/lib/marked.esm.js",
    "oak/": "https://deno.land/x/oak@v9.0.1/",
    "preact": "https://ga.jspm.io/npm:preact@10.6.4/dist/preact.mjs",
    "preact/compat": "https://ga.jspm.io/npm:preact@10.6.4/compat/dist/compat.mjs",
    "preact/debug": "https://ga.jspm.io/npm:preact@10.6.4/debug/dist/debug.mjs",
    "preact/devtools": "https://ga.jspm.io/npm:preact@10.6.4/devtools/dist/devtools.mjs",
    "preact/hooks": "https://ga.jspm.io/npm:preact@10.6.4/hooks/dist/hooks.mjs",
    "preact-render-to-string": "https://ga.jspm.io/npm:preact-render-to-string@5.1.19/dist/index.mjs",
    "prismjs": "https://ga.jspm.io/npm:prismjs@1.25.0/prism.js",
    "react": "https://ga.jspm.io/npm:preact@10.6.4/compat/dist/compat.mjs",
    "react-dom": "https://ga.jspm.io/npm:preact@10.6.4/compat/dist/compat.mjs",
    "react-dom/server": "https://ga.jspm.io/npm:preact@10.6.4/compat/server.mjs",
    "react-waterfall-render/": "https://ga.jspm.io/npm:react-waterfall-render@3.0.0/",
    "ruck/": "http://localhost:3002/",
    "std/": "https://deno.land/std@0.117.0/"
  }
}
```

Here is attempt 1:

<img width="899" alt="Screen Shot 2021-12-14 at 10 48 17 pm" src="https://user-images.githubusercontent.com/1754873/145993019-deb55304-0ffd-4034-bf65-d5b74363ea51.png">

And attempt 2:

<img width="885" alt="Screen Shot 2021-12-14 at 10 47 43 pm" src="https://user-images.githubusercontent.com/1754873/145993083-27bd5260-1e0c-4b8f-a679-665aa087200c.png">

1. Why do the `/` in the names get converted to `-`?
2. How come the import map still contains published CDN URLs?
3. Why would this ever work. For it to make sense, there would need to be a build directory, with the local preact converted like the CDN does; the source minified, and dev and prod versions of files created, etc.

**Comment by guybedford** [comment]
Here's how to do it:

```js
import { Generator } from '@jspm/generator';

const generator = new Generator({
  env: ['production', 'browser', 'module'],
  resolutions: {
    preact: './preact'
  }
});

await generator.traceInstall('./app.js');
console.log(generator.getMap());
```

Where eg `app.js` contains `import 'preact'` or imports other libraries that import preact etc.

Note that resolutions are like dependencies - there is no need to define them for separate subpaths - eg you wouldn't do `"dependencies": { "preact/compat": "1.2.3" }` in npm.

**Comment by guybedford** [comment]
For Preact to import, you just need to build it locally so the module version of Preact is available.

**Comment by jaydenseric** [comment]
We need to replicate the exact behavior of the live JSPM CDN when importing Preact. It minifies the files, and creates seperate dev and prod files depending on the contents, etc.

So for your suggested solution, what directory does the locally cloned Preact package get built to? I can't see anywhere you are specifying it.

**Comment by guybedford** [comment]
@jaydenseric Preact has its own build steps which are run before it is published to npm, you need to run `npm install && npm run build` in the cloned Preact repo - there is no way around that.

**Comment by jaydenseric** [comment]
I don't think you understand. The local `preact` directory is exactly the unzipped results of `npm pack` in the Preact repo clone. It's what would be published to npm. It's the result of all the building that normally happens when Preact is published.

The problem is, the JSPM CDN does modifications to files itself. It minifies them, it creates seperate dev and prod versions, etc. How can we get that to happen locally? It's already a compromise this has to happen locally; ideally you could deploy and test staging sites or even deploy to prod using a dependency fork. But anyway.

**Comment by jaydenseric** [comment]
For this:

https://generator.jspm.io/#a+NhYGBkDM0rySzJSU1hSC9KLMgozNEtSk1MLnEwNNMz0DPQT0stSc5wB8kE+ujlZhUDAONE0pI0AA

Compare what is published to npm:

https://unpkg.com/browse/graphql-react@16.0.0/fetchGraphQL.mjs

To what the JSPM CDN serves:

https://ga.jspm.io/npm:graphql-react@16.0.0/fetchGraphQL.mjs

**Comment by guybedford** [comment]
@jaydenseric I've mentioned this before but please do try work on your tone a little - "I don't think you understand" as the first sentence of a comment isn't ideal - you will get the best support if you avoid anything that could be misconstrued as ad hominem.

Any package which is shipped as an ES module and follows native semantics with a good build is fine to use directly in serving and in the generator. Preact is a great example of one of those packages where the JSPM optimizations are not necessary since Preact already does good optimizations itself on publish. Thus you can get this workflow working for Preact just fine using the steps I've provided above. If you are seeing any further issues with this please let me know.

Therefore I don't think you have a Preact problem.

For packages where you want minification and optimization of entry points, it's relatively straight forward to set this optimization up with RollupJS and Terser.

For CommonJS or more complex optimizations, the JSPM package builder will be open sourced to allow this eventually, it's just not a high priority for the project.

**Comment by jaydenseric** [comment]
Sorry for the tone, you can always give me the benefit of the doubt because I have the utmost respect for you and your work; you're one of my favourite people. I said that you're not understanding the issue in good faith and I didn't mean it as an attack but literally as we were 12 comments in and it turns out the issue really has a straight up answer: The live JSPM CDN doesn't have a way to specify a GitHub branch or commit hash instead of an npm package version, and there is no local solution for recreating the exact JSPM CDN behavior as the JSPM package builder isn't open sourced yet.

This issue is not just about how to try a `preact` fork; it's a general question that applies to forks of any package on npm, many of which are CJS and require the JSPM package builder magic to function in an ESM project. Even though `preact` distributes ESM files, I'm measuring lighthouse scores and testing performance with throttled network speeds and things where we really need to replicate exactly the prod JSPM CDN behavior that serve minified modules. I also want to see how the real JSPM CDN would pull in `process` polyfills, break on certain uses of `process.env.NODE_ENV`, etc. so there is no surprises in prod once `preact` v11 ships. I wanted to provide feedback to `preact` on how to optimise v11 for the JSPM CDN _before_ it's published.

**Comment by guybedford** [comment]
Right, so the package builder will be open sourced, and I even previously agreed to give someone access to the code to do the work of separating it from the monolith themselves as a collaboration. It may or may not be under a non-CDN license (eg GNU Affero or some such), or it may be completely permissive I haven't decided.

Another interesting option would be to have eg a CI action which uses the live CDN to run a build of a custom resource such as a GitHub repo or git package and provides the built CDN URL. The CDN does namespace `npm:` specifically to allow other namespaces like `github:` in this way. This way eg a CI system could run the JSPM CDN build, get a hook callback, then test against that.

**Comment by jaydenseric** [comment]
It's possible for the live JSPM CDN to be enhanced to allow repo URLs to be used in place of npm package versions, in the same way [`package.json` dependency versions can be specified as repo URLs](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#git-urls-as-dependencies) instead of a semver string. If the package has a build step, it must have the conventional `prepare` or other standard lifecycle script so tools (e.g. the [`npm pack` command](https://docs.npmjs.com/cli/v8/using-npm/scripts#npm-pack)) can prepare the package before packing it, to get the "published" files. The JSPM CDN builder could then work on the resulting tarball.
